crates/compiler/can/src/def.rs

```diff
@@ -2880,6 +2941,188 @@ fn to_pending_value_def<'a>(
+                                    .map(|(_, ident)| Lowercase::from(ident))
+                                    .collect();
+
+                                env.problem(Problem::RuntimeError(RuntimeError::ValueNotExposed {
+                                    module_name: module_name.clone(),
+                                    ident,
+                                    region: loc_name.region,
+                                    exposed_values,
+                                }))
+                            }
+                        }
+                    }
+                }
+            }
+
+            PendingValue::ModuleImport(IntroducedImport {
+                module_id,
+                region,
+                exposed_symbols,
+            })
+        }
+        IngestedFileImport(ingested_file) => {
+            let file_path =
+                if let ast::StrLiteral::PlainLine(ingested_path) = ingested_file.path.value {
+                    let mut file_path: PathBuf = env.module_path.into();
+                    // Remove the header file name and push the new path.
+                    file_path.pop();
+                    file_path.push(ingested_path);
+
+                    match fs::metadata(&file_path) {
+                        Ok(md) => {
+                            if md.is_dir() {
+                                env.problem(Problem::FileProblem {
+                                    filename: file_path.clone(),
+                                    // TODO: change to IsADirectory once that is stable.
+                                    error: io::ErrorKind::InvalidInput,
+                                });
+                                return PendingValue::InvalidIngestedFile;
+                            }
+                            file_path
+                        }
+                        Err(e) => {
+                            env.problem(Problem::FileProblem {
+                                filename: file_path,
+                                error: e.kind(),
+                            });
+                            return PendingValue::InvalidIngestedFile;
+                        }
+                    }
+                } else {
+                    todo!(
+                    "Only plain strings are supported. Other cases should be made impossible here"
+                );
+                };
+
+            let typed_ident = ingested_file.name.item.extract_spaces().item;
+            let body_pattern = env
+                .arena
+                .alloc(typed_ident.ident.map_owned(|ident|
+                        ast::Pattern::Identifier {
+                            ident,
+                            suffixed: 0
+                        }
+                    ));
+            let ann_type = env.arena.alloc(typed_ident.ann);
+            let body_expr = Loc {
+                value: ast::Expr::IngestedFile(env.arena.alloc(file_path), ann_type),
+                region,
+            };
```

Comment:
It feels strange to me that the parsed AST is being constructed here. Is there a reason this can't be canonicalized directly rather than indirecting through the AST again?

Reply:
Totally. I copied this from `load` where it made sense because the ingested file was in the header, so it had to construct an `Expr` for `can` later, but it's definitely unnecessary now that we are there already.

Reply:
I think improving that can be in a separate PR! 😃

Reply:
Oh, didn't see this comment. It's already refactored [here](https://github.com/roc-lang/roc/pull/6658/commits/7efc7a4a3e396670a4458cc3a236c1f2e5bb5d01).